ui/summary.go

```diff
@@ -287,73 +361,78 @@ func SummarizeMetrics(w io.Writer, indent string, t time.Duration, timeUnit stri
-				}
-			}
+			updateExtraMaxLens(extra, extraMaxLens)
 		}
 	}
 
 	sort.Strings(names)
-	tmpCols := make([]string, len(TrendColumns))
+
 	for _, name := range names {
 		m := metrics[name]
+		mark, markColor := metricsMark(m)
+		fmtName := displayNameForMetric(m)
+		fmtIndent := indentForMetric(m)
+		fmtName += GrayColor.Sprint(strings.Repeat(".", nameLenMax-StrWidth(fmtName)-StrWidth(fmtIndent)+3) + ":")
+		fmtData := formatData(name, trendCols[name], trendColMaxLens, values, valueMaxLen, extras[name], extraMaxLens)
+		_, _ = fmt.Fprint(w, indent+fmtIndent+markColor.Sprint(mark)+" "+fmtName+" "+fmtData+"\n")
+	}
+}
 
-		mark := " "
-		markColor := StdColor
-		if m.Tainted.Valid {
-			if m.Tainted.Bool {
-				mark = FailMark
-				markColor = FailColor
-			} else {
-				mark = SuccMark
-				markColor = SuccColor
-			}
-		}
+// Summarize summarizes a dataset in human readable format.
+func Summarize(w io.Writer, indent string, data SummaryData) {
+	if data.Root != nil {
+		summarizeGroup(w, indent+"    ", data.Root)
+	}
+	summarizeMetrics(w, indent+"  ", data.Time, data.Opts.SummaryTimeUnit.String, data.Metrics)
+}
 
-		fmtName := DisplayNameForMetric(m)
-		fmtIndent := IndentForMetric(m)
-		fmtName += GrayColor.Sprint(strings.Repeat(".", nameLenMax-StrWidth(fmtName)-StrWidth(fmtIndent)+3) + ":")
+func newJSONEncoder(w io.Writer) *json.Encoder {
+	encoder := json.NewEncoder(w)
+	encoder.SetIndent("", "    ")
+	return encoder
+}
+func summarizeGroupJSON(w io.Writer, group *lib.Group) {
+	encoder := newJSONEncoder(w)
+	if err := encoder.Encode(group); err != nil {
+		panic(err)
+	}
+}
 
-		var fmtData string
-		if cols := trendCols[name]; cols != nil {
-			for i, val := range cols {
-				tmpCols[i] = TrendColumns[i].Key + "=" + ValueColor.Sprint(val) + strings.Repeat(" ", trendColMaxLens[i]-StrWidth(val))
-			}
-			fmtData = strings.Join(tmpCols, " ")
-		} else {
-			value := values[name]
-			fmtData = ValueColor.Sprint(value) + strings.Repeat(" ", valueMaxLen-StrWidth(value))
-
-			extra := extras[name]
-			switch len(extra) {
-			case 0:
-			case 1:
-				fmtData = fmtData + " " + ExtraColor.Sprint(extra[0])
-			default:
-				parts := make([]string, len(extra))
-				for i, ex := range extra {
-					parts[i] = ExtraColor.Sprint(ex) + strings.Repeat(" ", extraMaxLens[i]-StrWidth(ex))
-				}
-				fmtData = fmtData + " " + ExtraColor.Sprint(strings.Join(parts, " "))
-			}
+func summarizeMetricsJSON(w io.Writer, t time.Duration, timeUnit string, metrics map[string]*stats.Metric) {
+	data := make(map[string]interface{})
+	for name, m := range metrics {
+		m.Sink.Calc()
+		if sink, ok := m.Sink.(*stats.TrendSink); ok {
+			data[name] = sink.Format(t)
+			continue
+		}
+		data[name] = m.Sink.Format(t)
+		_, extra := nonTrendMetricValueForSum(t, timeUnit, m)
+		if len(extra) > 1 {
+			extraData := make(map[string]interface{})
+			extraData["value"] = m.Sink.Format(t)["value"]
+			extraData["extra"] = extra
+			data[name] = extraData
```

Comment:
This seems ... strange to me ... I guess I will have to get a lot better look at the whole code (both the for the cli summary and the stats.Sink#Format), but it seems like we are doing very similar things in different places and than we discard half the things we did and do the again ... but slightly differently ... maybe

Reply:
@MStoykov I think the only part that relevant is in summaryMetrics:

```
		extras[name] = extra
		if len(extra) > 1 {
			for i, ex := range extra {
				if l := StrWidth(ex); l > extraMaxLens[i] {
					extraMaxLens[i] = l
				}
			}
		}
```